xnmt/inference.py

```diff
@@ -72,12 +72,30 @@ def __call__(self, generator, src_file=None, trg_file=None, candidate_id_file=No
     # Corpus
     src_corpus = list(generator.src_reader.read_sents(args["src_file"]))
     # Get reference if it exists and is necessary
-    if args["mode"] == "forced" or args["mode"] == "forceddebug":
+    if args["mode"] == "forced" or args["mode"] == "forceddebug" or args["mode"] == "score":
       if args["ref_file"] == None:
         raise RuntimeError("When performing {} decoding, must specify reference file".format(args["mode"]))
-      ref_corpus = list(generator.trg_reader.read_sents(args["ref_file"]))
-      if self.max_len and any(len(s)>self.max_len for s in ref_corpus):
-        logger.warning("Forced decoding with some targets being longer than max_len. Increase max_len to avoid unexpected behavior.")
+      if args["mode"] == "score":
```

Comment:
This is indeed a bit hacky. What about (mostly) deleting this if statement, and then stepping through the n-best list one line at a time below?